src/opentimelineio/clip.cpp

```diff
@@ -19,33 +23,117 @@ Clip::~Clip()
+               : active->second;
+}
+
+Clip::MediaReferences
+Clip::media_references() const noexcept
+{
+    MediaReferences result;
+    for (auto const& m: _media_references)
+    {
+        result.insert(
+            { m.first, dynamic_retainer_cast<MediaReference>(m.second) });
+    }
+    return result;
+}
+
+template <typename MediaRefMap>
+bool
+Clip::check_for_valid_media_reference_key(
+    std::string const& key,
+    MediaRefMap const& media_references,
+    ErrorStatus*       error_status)
+{
+    auto found = media_references.find(key);
+    if (found == media_references.end())
+    {
+        if (error_status)
+        {
+            *error_status = ErrorStatus(
+                ErrorStatus::MEDIA_REFERENCES_MISSING_ACTIVE_KEY,
+                "The media references do not contain the active key",
+                this);
+        }
+        return false;
+    }
+    return true;
+}
+
+void
+Clip::set_media_references(
+    MediaReferences const& media_references,
+    std::string const&     new_active_key,
+    ErrorStatus*           error_status) noexcept
+{
+    if (!check_for_valid_media_reference_key(
+            new_active_key, media_references, error_status))
+    {
+        return;
+    }
+
+    _media_references.clear();
+    for (auto const& m: media_references)
+    {
+        _media_references[m.first] = m.second ? m.second : new MissingReference;
+    }
+
+    if (!new_active_key.empty())
+    {
+        _active_media_reference_key = new_active_key;
+    }
+}
+
+std::string
+Clip::active_media_reference_key() const noexcept
+{
+    return _active_media_reference_key;
+}
+
+void
+Clip::set_active_media_reference_key(
+    std::string const& new_active_key, ErrorStatus* error_status) noexcept
+{
+    if (!check_for_valid_media_reference_key(
+            new_active_key, _media_references, error_status))
+    {
+        return;
+    }
+    _active_media_reference_key = new_active_key;
 }
 
 void
 Clip::set_media_reference(MediaReference* media_reference)
 {
-    _media_reference = media_reference ? media_reference : new MissingReference;
+    _media_references[_active_media_reference_key] =
+        media_reference ? media_reference : new MissingReference;
 }
 
 bool
 Clip::read_from(Reader& reader)
 {
-    return reader.read("media_reference", &_media_reference) &&
+    return reader.read("media_references", &_media_references) &&
+           reader.read(
+               "active_media_reference_key", &_active_media_reference_key) &&
            Parent::read_from(reader);
 }
 
 void
 Clip::write_to(Writer& writer) const
 {
     Parent::write_to(writer);
-    writer.write("media_reference", _media_reference);
+    writer.write("media_references", _media_references);
+    writer.write("active_media_reference_key", _active_media_reference_key);
 }
 
 TimeRange
 Clip::available_range(ErrorStatus* error_status) const
 {
-    if (!_media_reference)
+    auto active_media = media_reference();
+    if (!active_media)
```

Comment:
does a special MissingReference check need to be done here or is it accommodated sufficiently by available_range below?

Reply:
A MissingReference will give an unset optional<TimeRange> in C++ or a None in Python. 

This is the same as it would have done before with a MissingReference as the sole media reference, but we can take this opportunity to do something better if you have a better suggestion.

Reply:
sounds good to me.